tests/test_json_parser.py

```diff
@@ -0,0 +1,75 @@
+# from .checks import check_extract_article
+from bs4 import BeautifulSoup
+from ..readabilipy.json_parser import plain_element, plain_text_leaf_node, add_node_indexes, content_digest
+from ..readabilipy.text_manipulation import normalise_text
+
+
+def test_plain_element_with_comments():
+    """Contents of comments should be stripped but the comment itself should be kept."""
+    html = """
+        <div>
+            <p>Text</p>
+            <!-- comment -->
+        </div>
+    """.strip()
+    soup = BeautifulSoup(html, 'html.parser')
+    elements = [str(plain_element(element, False, False)) for element in soup.contents]
+    assert elements == ["<div><p>Text</p><!----></div>"]
+
+
+def test_content_digest_on_filled_and_empty_elements():
+    """Filled strings should get a digest but empty strings should not."""
+    html = """
+        <div>
+            <p>Text</p>
+            <p></p>
+        </div>
+    """.strip()
+    soup = BeautifulSoup(html, 'html.parser')
+    elements = [str(plain_element(element, True, True)) for element in soup.contents]
+    assert elements == ['<div><p data-content-digest="71988c4d8e0803ba4519f0b2864c1331c14a1890bf8694e251379177bfedb5c3">Text</p><p data-content-digest=""></p></div>']
+
+
+def test_leaf_nodes_without_text():
+    """Leaf nodes with text should yield their text, while those without should yield None."""
```

Comment:
See above comment about empty strings. I think this is probably the right behaviour in terms of not amending the structure when "plaintextifying" the simplified HTML. However, I could potentially be persuaded that we should remove paragraphs that would be empty in the plain text.